File: java/src/org/revenj/patterns/Container.java
package org.revenj.patterns;

public interface Container extends ServiceLocator, AutoCloseable {
	interface Factory<T> {
		T create(Container scope);
	}

	<T> void register(Class<T> manifest, boolean singleton);

	<TInterface, TService extends TInterface> void register(Class<TService> manifest, Class<TInterface> as, boolean singleton);

	<T> void register(boolean closable, T service);

	default <T> void register(T service) {
		register(service instanceof AutoCloseable, service);
	}

	<TInterface, TService extends TInterface> void register(TService service, Class<TInterface> as);

	<T> void register(Class<T> manifest, boolean singleton, Factory<T> service);

	Container createScope();
}

File: scala/revenj-core/src/net/revenj/extensibility/Container.scala
package net.revenj.extensibility

import net.revenj.patterns.ServiceLocator

import scala.reflect.runtime.universe.TypeTag

trait Container extends ServiceLocator {
  def registerClass[T: TypeTag](manifest: Class[T], singleton: Boolean)

  def registerInstance[T: TypeTag](service: T, handleClose: Boolean)

  def registerFactory[T: TypeTag](factory: Container => T, singleton: Singleton)

  def createScope(): Container
}
